### PR TITLE
ServeDir & ServeFile last modified headers handling

### PR DESCRIPTION
**Pull request by zenria**
~Note: This PR also includes https://github.com/tower-rs/tower-http/pull/137, I will modify it when it will be merged~

## Motivation

ServeDir & ServeFile should respect http caching best practice by setting `Last-Modified` response header and handling `If-Modified_since` & `If-Unmodified-Since` requests headers.

## Solution

The `Last-Modified` header is added on responses of ServeFile & ServeDir. 

`If-Modified_since` & `If-Unmodified-Since` requests headers are handled so the service will respond `NOT_MODIFIED` or `PRECONDITION_FAILED` depending on values set into the headers.

### CODE REVIEW

**Comment by davidpdrsn** [comment]
@zenria do you merge/rebase the latest master? Seems some things have changed making this hard to review.

**Comment by zenria** [comment]
> @zenria do you merge/rebase the latest master? Seems some things have changed making this hard to review.

I've just rebased my branch on master :)

**Comment by davidpdrsn** [comment]
Thank you! Given the size of this change I'm pushing it to 0.2 as I would like to get 0.1.2 our relatively soon.

File: tower-http/Cargo.toml
Comment (davidpdrsn) [inline tower-http/Cargo.toml:37]: I'm not sure we should depend on `headers`. Using the stuff in `http::headers` is fine imo. Not worth adding another dependency for.

**Review by davidpdrsn — CHANGES_REQUESTED** [review]

**Comment by davidpdrsn** [comment]
We made some more changes to CI end of last week so you'll have to rebase `master` again to get CI working.

Just a heads up there are also some other PRs landing this week which touch these files so you'll probably get some merge conflicts soon.

File: tower-http/Cargo.toml
Comment (zenria) [reply tower-http/Cargo.toml:37]: I'll have a look on how to remove this dependency

**Comment by zenria** [comment]
I've rebased my branch, you were right, there was some merge conflicts ;)

I've got rid of `headers` crate, but I was forced to introduce a dependency on `httpdate` to parse and generate headers.

**Comment by davidpdrsn** [comment]
Thanks! I'll take a look some time this week.

**Comment by davidpdrsn** [comment]
Just a heads up. I'm waiting on https://github.com/tower-rs/tower-http/pull/173 being merged before diving into this.

**Comment by davidpdrsn** [comment]
Removing this from the 0.2 milestone. It wont require breaking changes so we can always ship it in 0.2.1

**Comment by shepmaster** [comment]
Thank you for your work on this; I'm excited to have this functionality in tower-http!

File: tower-http/src/services/fs/serve_dir.rs
Comment (davidpdrsn) [inline tower-http/src/services/fs/serve_dir.rs:271]: This should use [`tokio::fs::metadata`](https://docs.rs/tokio/latest/tokio/fs/fn.metadata.html) to not block the runtime. We already do some things with the metadata on `master` so if you rebase it would be relatively easy to fix.

**Review by davidpdrsn — CHANGES_REQUESTED** [review]
Looks good overall! Using `httpdate` for parsing is fine. Hyper is also using it.

I'm sorry but it looks like you'll have to rebase `master` again 😞 At least `ServeFile` has been simplified quite a bit which should make this simpler as well.

**Comment by zenria** [comment]
I've rebased my branch :)

File: tower-http/src/services/fs/mod.rs
Comment (shepmaster) [inline tower-http/src/services/fs/mod.rs:239]: `map` followed by `flatten` should be `flat_map`

File: tower-http/src/services/fs/mod.rs
Comment (shepmaster) [inline tower-http/src/services/fs/mod.rs:256]: Ditto

File: tower-http/src/services/fs/mod.rs
Comment (shepmaster) [inline tower-http/src/services/fs/mod.rs:262]: Usually you don’t want references to options, but an optional reference. `Option::as_ref` Can be used

File: tower-http/src/services/fs/serve_dir.rs
Comment (shepmaster) [inline tower-http/src/services/fs/serve_dir.rs:319]: Ditto here x2

File: tower-http/src/services/fs/serve_dir.rs
Comment (shepmaster) [inline tower-http/src/services/fs/serve_dir.rs:462]: Feels unfortunate that this went from one possibility to hundreds when it seemingly only needed two possibilities.

File: tower-http/src/services/fs/serve_dir.rs
Comment (shepmaster) [inline tower-http/src/services/fs/serve_dir.rs:752]: Why’s this get removed?

File: tower-http/src/services/fs/serve_file.rs
Comment (shepmaster) [inline tower-http/src/services/fs/serve_file.rs:1]: Why’s this get removed?

**Review by shepmaster — COMMENTED** [review]
My comments don’t carry any weight in this repository, just actively reading through to see what changed.

File: tower-http/src/services/fs/serve_dir.rs
Comment (zenria) [reply tower-http/src/services/fs/serve_dir.rs:752]: Sorry, this is a merge error!

File: tower-http/src/services/fs/serve_file.rs
Comment (zenria) [reply tower-http/src/services/fs/serve_file.rs:1]: this too :(

File: tower-http/src/services/fs/serve_dir.rs
Comment (zenria) [reply tower-http/src/services/fs/serve_dir.rs:462]: Yes, you're right!

File: tower-http/src/services/fs/mod.rs
Comment (zenria) [reply tower-http/src/services/fs/mod.rs:239]: yes! you're right. (btw this is an option, as there is no `flat_map` on that type, I always forget I can use `and_then` to do that.)

File: tower-http/src/services/fs/mod.rs
Comment (zenria) [reply tower-http/src/services/fs/mod.rs]: in that case this does not matter, but I agree.

File: tower-http/src/services/fs/serve_dir.rs
Comment (zenria) [reply tower-http/src/services/fs/serve_dir.rs:460]: the `check_modified_headers` can output 2 different status codes: `304 Not Modified` and `412 Precondition Failed` this is why I'm not using 2 more enum variants.

File: tower-http/src/services/fs/mod.rs
Comment (Nehliin) [inline tower-http/src/services/fs/mod.rs:255]:
```suggestion
        std::str::from_utf8(value.as_bytes()))
            .ok()
            .and_then(|value| httpdate::parse_http_date(value).ok())
            .map(|time| IfUnmodifiedSince(time.into()))
```

File: tower-http/src/services/fs/serve_file.rs
Comment (Nehliin) [inline tower-http/src/services/fs/serve_file.rs:472]: Could we please also check that the body is empty here like the spec states?

File: tower-http/src/services/fs/serve_file.rs
Comment (Nehliin) [inline tower-http/src/services/fs/serve_file.rs:481]: Let's check that the file is actually part of the body here as a sanity check.

File: tower-http/src/services/fs/serve_file.rs
Comment (Nehliin) [inline tower-http/src/services/fs/serve_file.rs:492]: Same here

File: tower-http/src/services/fs/serve_file.rs
Comment (Nehliin) [inline tower-http/src/services/fs/serve_file.rs:501]: And this shouldn't return the body.

File: tower-http/src/services/fs/serve_dir.rs
Comment (Nehliin) [inline tower-http/src/services/fs/serve_dir.rs:1164]: Same comments here like in the tests for serve_file

**Review by Nehliin — APPROVED** [review]
Nicely done! Thanks for the PR! I have some comments regarding the tests but those extra checks should still pass with this implementation. Also have a comment regarding some unnecessary heap allocations. 

I will approve this since I personally think my comments can be easily fixed in a follow up pr by either you or me and this has been in the works for some time now.  But it's up to @davidpdrsn if he wants to block this on my comments.

**Comment by zenria** [comment]
Ok, thank you @Nehliin I'll update the PR with the CI & suggested fixes :)

File: tower-http/src/services/fs/mod.rs
Comment (zenria) [reply tower-http/src/services/fs/mod.rs]: Ok, I was not aware of this function, thanks a lot!

**Review by davidpdrsn — APPROVED** [review]

**Comment by photino** [comment]
I cannot find the `Last-Modified` header in the response for `tower-http` 0.4.4.

**Comment by davidpdrsn** [comment]
Please open a new issue instead of resurrecting such and old PR.

**Pull request merged**
